applogic/src/api/user/user_cubit.rs

```diff
@@ -0,0 +1,174 @@
+
+use super::{StreamSink, User};
+
+/// Logged in user
+///
+/// Opaque, cheaply clonable, copy-on-write type
+///
+/// Note: This has a prefix `Ui` to avoid conflicts with the `User`.
+//
+// TODO: Currently, frb does not support exposing eq and hash to Dart. When it is possible, we
+// should do it, to minimize the amount of UI rebuilds in Flutter.
+//
+// See:
+// * <https://github.com/phnx-im/infra/issues/247>
+// * <https://github.com/fzyzcjy/flutter_rust_bridge/issues/2238>
+#[frb(opaque)]
+#[derive(Debug, Clone)]
+pub struct UiUser {
+    inner: Arc<UiUserInner>,
+}
+
+#[derive(Debug)]
+struct UiUserInner {
+    user_name: QualifiedUserName,
+    profile: Option<UserProfile>,
+}
+
+impl UiUser {
+    fn new(user_name: QualifiedUserName, profile: Option<UserProfile>) -> Self {
+        let inner = Arc::new(UiUserInner { user_name, profile });
+        Self { inner }
+    }
+
+    /// Loads the user profile in the background
+    fn spawn_load(this: Arc<RwLock<Self>>, core_user: CoreUser) {
+        spawn_from_sync(async move {
+            match core_user.own_user_profile().await {
+                Ok(profile) => {
+                    let mut state = this.write().await;
+                    *state = UiUser::new(state.inner.user_name.clone(), Some(profile));
+                }
+                Err(error) => {
+                    log::error!("Could not load own user profile: {:?}", error);
+                }
+            }
+        });
+    }
+
+    #[frb(getter, sync)]
+    pub fn user_name(&self) -> String {
+        self.inner.user_name.to_string()
+    }
+
+    #[frb(getter, sync)]
+    pub fn display_name(&self) -> Option<String> {
+        let profile = self.inner.profile.as_ref()?;
+        Some(profile.display_name()?.to_string())
+    }
+
+    #[frb(getter, sync)]
+    pub fn profile_picture(&self) -> Option<Vec<u8>> {
+        let profile = self.inner.profile.as_ref()?;
+        Some(profile.profile_picture()?.value()?.to_vec())
+    }
+}
+
+/// Provides access to the logged in user and their profile.
+///
+/// Note: this has a suffix `Base` because the corresponding Dart class does not implement
+/// `StateStreamableSource`, and therefore to impemlement it we need to wrap it Dart.
+#[frb(opaque)]
+pub struct UserCubitBase {
+    state: Arc<RwLock<UiUser>>,
+    sinks: Option<Vec<StreamSink<UiUser>>>,
+    core_user: CoreUser,
+}
+
+impl UserCubitBase {
+    #[frb(sync)]
+    pub fn new(user: &User) -> Self {
+        info!("UserCubitBase::new");
+
+        let core_user = user.user.clone();
+        let state = Arc::new(RwLock::new(UiUser::new(core_user.user_name(), None)));
+
+        UiUser::spawn_load(state.clone(), core_user.clone());
+
+        // TODO: We are missing subscription to changes of the user from the core user. Since,
+        // there is no way to change the user profile from other place than this cubit, it is not a
+        // problem currently. However, in a multi-client scenario, this will become one.
```

Comment:
Please also file an issue for these internal TODOs, that'll make project management easier going forward.

Reply:
Also, regarding this specific topic, this is exactly where we need a mechanism to inform cubits of changes of the underlying data. In that sense it's not an isolated issue but rather something we'll need everywhere.

Reply:
Indeed. I changed the comment and added a reference to https://github.com/phnx-im/infra/issues/254.